### PR TITLE
change(web): element wrapper subpackage modularization 🧩

### PR DESCRIPTION
**Pull request by jahorton**

### CODE REVIEW

File: web/src/engine/element-wrappers/src/input.ts [hunk 1]
```diff
@@ -0,0 +1,233 @@
...
+      // TODO:  rework class to accept a class-static 'callback' from the DOM module that this can call.
+      //        Would eliminate the need for this 'static' reference.
+      //        Only strongly matters once we better modularize KMW, with web-dom vs web-dom-targets vs web-core, etc.
+      if(com.keyman["singleton"]) {
+        com.keyman["singleton"].domManager.moveToNext(false);
+      }
+   ```
+   * This does not belong in a modularized version of this class; it must be supplied
+   * by the consuming top-level products instead.
+   */
```
Comment (mcdurdin) [inline web/src/engine/element-wrappers/src/input.ts:53]: TODO: complete modularization in future PR

File: web/src/engine/element-wrappers/src/input.ts [hunk 2]
```diff
@@ -0,0 +1,233 @@
...
+    let domStart = this.root.value._kmwCodePointToCodeUnit(start);
+    let domEnd = this.root.value._kmwCodePointToCodeUnit(end);
+    this.root.setSelectionRange(domStart, domEnd, direction);
+
+    this.processedSelectionStart = start;
+    this.processedSelectionEnd = end;
+
+    this.events.emit('scrollfocusrequest', this.root);
+
+    this.root.setSelectionRange(domStart, domEnd, direction);
```
Comment (mcdurdin) [inline web/src/engine/element-wrappers/src/input.ts:151]: Noting that this replaces `Utils.forceScroll(this.root);`

File: web/src/engine/element-wrappers/src/input.ts [hunk 3]
```diff
@@ -0,0 +1,233 @@
...
+  }
+
+  handleNewlineAtCaret(): void {
+    const inputEle = this.root;
+    // Can't occur for Mocks - just Input types.
+    if (inputEle && (inputEle.type == 'search' || inputEle.type == 'submit')) {
+      inputEle.disabled=false;
+      inputEle.form.submit();
+    } else {
+      this.events.emit('unhandlednewline', inputEle);
```
Comment (mcdurdin) [inline web/src/engine/element-wrappers/src/input.ts:226]: Noting that this replaces

```
if(com.keyman["singleton"]) {
  com.keyman["singleton"].domManager.moveToNext(false);
}
```

File: web/src/engine/element-wrappers/src/outputTarget.ts [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+    // This class has non-integrated unit tests in which the `singleton` object doesn't exist.
+    // Thus, we need to test for this case.
+    let keyman = com.keyman['singleton'];
+
+    // Signal the necessary text changes to the embedding app, if it exists.
+    if(keyman && keyman['oninserttext'] && keyman.isEmbedded) {
+      keyman['oninserttext'](transform.deleteLeft, transform.insert, transform.deleteRight);
+    }
+   ```
+   */
```
Comment (mcdurdin) [inline web/src/engine/element-wrappers/src/outputTarget.ts:17]: TODO: complete modularization

File: web/src/engine/element-wrappers/src/textarea.ts [hunk 5]
```diff
@@ -0,0 +1,216 @@
...
+    element.selectionStart = selectionStart;
+    element.selectionEnd = selectionEnd;
+  }
+}
+   ```
+   * References to the event-handlers & related states objects are not available within this submodule.
+   *
+   * It is the parts between and including the _IgnoreBlurFocus references that must be
+   * implemented externally.
+   */
```
Comment (mcdurdin) [inline web/src/engine/element-wrappers/src/textarea.ts:34]: TODO: complete modularization

File: web/src/engine/element-wrappers/src/textarea.ts [hunk 6]
```diff
@@ -0,0 +1,216 @@
...
+    let domStart = this.root.value._kmwCodePointToCodeUnit(start);
+    let domEnd = this.root.value._kmwCodePointToCodeUnit(end);
+    this.root.setSelectionRange(domStart, domEnd, direction);
+
+    this.processedSelectionStart = start;
+    this.processedSelectionEnd = end;
+
+    this.events.emit('scrollfocusrequest', this.root);
+
+    this.root.setSelectionRange(domStart, domEnd, direction);
```
Comment (mcdurdin) [inline web/src/engine/element-wrappers/src/textarea.ts:141]: Note, replaces `Utils.forceScroll(this.root);`

File: web/src/test/auto/cases/attachmentAPI.js [hunk 7]
```diff
@@ -1,257 +1,257 @@
...
-    it('contentEditable=true', function(done) {
-      var ID = DynamicElements.addEditable();
-      var ele = document.getElementById(ID);
-
-      DynamicElements.assertAttached(ele, done);
-    });
-  });
-});
\ No newline at end of file
+// var assert = chai.assert;
```
Comment (mcdurdin) [inline web/src/test/auto/cases/attachmentAPI.js:1]: TODO: fix these

File: web/src/test/auto/cases/basics.js [hunk 8]
```diff
@@ -1,154 +1,154 @@
...
-      })
-    });
-
-    describe('Basic Toolbar UI', function() {
-
-      beforeEach(function() {
-        this.timeout(testconfig.timeouts.uiLoad);
-        fixture.setBase('fixtures');
-        fixture.load('singleInput.html');
+// var assert = chai.assert;
```
Comment (mcdurdin) [inline web/src/test/auto/cases/basics.js:1]: TODO: fix these

File: web/src/test/auto/cases/engine.js [hunk 9]
```diff
@@ -1,292 +1,292 @@
...
-        assert.isNotNull(keyman.getKeyboard("lao_2008_basic", "lo"), "Keyboard stub was not registered!");
-        assert.equal(keyman.getActiveKeyboard(), "Keyboard_lao_2008_basic", "Keyboard not set automatically!");
-        keyman.removeKeyboards('lao_2008_basic');
-        assert.equal(keyman.getActiveKeyboard(), '', "Keyboard not removed correctly!");
-      });  // THEN we run our checks.
-    });
-  });
-});
\ No newline at end of file
+// var assert = chai.assert;
```
Comment (mcdurdin) [inline web/src/test/auto/cases/engine.js:1]: TODO: fix these

File: web/src/test/auto/cases/engine_chirality.js [hunk 10]
```diff
@@ -1,47 +1,47 @@
-var assert = chai.assert;
+// var assert = chai.assert;
```
Comment (mcdurdin) [inline web/src/test/auto/cases/engine_chirality.js:1]: TODO: fix these

File: web/src/test/auto/cases/events.js [hunk 11]
```diff
@@ -1,113 +1,113 @@
-var assert = chai.assert;
+// var assert = chai.assert;
```
Comment (mcdurdin) [inline web/src/test/auto/cases/events.js:1]: TODO: fix these

File: web/src/test/auto/cases/text_selection.js [hunk 12]
```diff
@@ -1,545 +1,545 @@
...
-          assert.deepEqual([ele.selectionStart, ele.selectionEnd], [2, 6], "Sanity check: expected selection to be '𐌁𐌂'");
-
-          await assertInputSteps(ele, 1, () => {
-            eventDriver.simulateEvent(keys.O);
-          });
-
-          assert.strictEqual(ele.value, '𐌀𐍈𐌈', "Expected selection to have been replaced with '𐍈'");
-          assert.deepEqual([ele.selectionStart, ele.selectionEnd], [4, 4], "Expected caret to be after '𐍈'");
-        });
+// var assert = chai.assert;
```
Comment (mcdurdin) [inline web/src/test/auto/cases/text_selection.js:1]: TODO: fix these

File: web/src/test/auto/test_utils.js [hunk 13]
```diff
@@ -1,200 +1,199 @@
...
-  });
-}
-
-function retrieveAndReset(Pelem) {
-  let val = Pelem.value;
-  Pelem.value = "";
-
-  return val;
-}
+// // Keyman test suite utility methods
```
Comment (mcdurdin) [inline web/src/test/auto/test_utils.js:16]: TODO: fix these

**Review by mcdurdin — APPROVED** [review]
Lots of TODO items here but apart from that LGTM

**Comment by jahorton** [comment]
For clarity, most of the latter "fix these" comments from mcdurdin above are about disabled unit test suites.  Those that are disabled simply aren't runnable at present because the parts being tested within them haven't been modularized yet.

**Pull request merged**